### ISSUE TITLE
IDE0004 incorrectly suggests a nullable value type cast is redundant

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#37951]
**Version Used**: 
3.3.0-beta3-19406-05+a1905991543bed104f7f7f0842aca2b65d263b87

**Steps to Reproduce**:
```C#
class Program
{
    public static void Main() { }

    public static bool? GetBool(string name) =>
        name switch
        {
            "a" => (bool?)true,
            "b" => false,
            _ => null,
        };
}
```

**Expected Behavior**:
IDE0004 does not trigger.

**Actual Behavior**:
IDE0004 triggers, suggesting that the `(bool?)` is redundant:
![image](https://user-images.githubusercontent.com/2642209/62952532-d2845c80-bdb9-11e9-8256-b8d1541693b5.png)
However, removing the cast then results in an error:
![image](https://user-images.githubusercontent.com/2642209/62952576-edef6780-bdb9-11e9-975a-b4eb7e6b3a5d.png)

**Comment by CyrusNajmabadi** [comment]
@dpoeschl was there a 'switch expression' test pass done for IDE features?

**Comment by ryzngard** [comment]
In 16.4P2 this still recommends removal of `(bool?)` but the code compiles correctly when returning 'bool?'. If there is ambiguity the removal is not suggested, such as in 

```c#
var s = "";
var b = s switch
            {
                "a" => (bool?)true,
                "b" => false,
                _ => null,
            };
```
